File: competitive-programming/leet-code/257/binaryTreePaths.js
var binaryTreePaths = function (root) {
  const res = [];
  if (!root) return res;

  const stack = [[root, `${root.val}`]];

  while (stack.length > 0) {
    const [node, path] = stack.pop();

    if (!node.left && !node.right) {
      res.push(path);
    }

    if (node.right) {
      stack.push([node.right, `${path}->${node.right.val}`]);
    }

    if (node.left) {
      stack.push([node.left, `${path}->${node.left.val}`]);
    }
  }

  return res;
};

File: competitive-programming/leet-code/257/binary_tree_paths.py
from typing import List, Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def binary_tree_paths(root: Optional[TreeNode]) -> List[str]:
    if not root:
        return []

    result = []
    stack = [(root, str(root.val))]

    while stack:
        node, path = stack.pop()

        if not node.left and not node.right:
            result.append(path)

        if node.right:
            stack.append((node.right, f"{path}->{node.right.val}"))

        if node.left:
            stack.append((node.left, f"{path}->{node.left.val}"))

    return result